app/forms/instance-create.tsx

```diff
@@ -573,24 +589,95 @@ export function CreateInstanceForm() {
   )
 }
 
+// `ip is …` guard is necessary until we upgrade to 5.5, which handles this automatically
+const isFloating = (
+  ip: ExternalIpCreate
+): ip is { type: 'floating'; floatingIp: NameOrId } => ip.type === 'floating'
+
+const FloatingIpLabel = ({ ip }: { ip: FloatingIp }) => (
+  <div>
+    <div>{ip.name}</div>
+    <div className="flex gap-0.5 text-tertiary selected:text-accent-secondary">
+      <div>{ip.ip}</div>
+      {ip.description && (
+        <>
+          <Slash />
+          <div className="grow overflow-hidden overflow-ellipsis whitespace-pre text-left">
+            {ip.description}
+          </div>
+        </>
+      )}
+    </div>
+  </div>
+)
+
 const AdvancedAccordion = ({
   control,
   isSubmitting,
+  floatingIpList,
   siloPools,
 }: {
   control: Control<InstanceCreateInput>
   isSubmitting: boolean
+  floatingIpList: { items: Array<FloatingIp> }
   siloPools: Array<{ name: string; isDefault: boolean }>
 }) => {
   // we track this state manually for the sole reason that we need to be able to
   // tell, inside AccordionItem, when an accordion is opened so we can scroll its
   // contents into view
   const [openItems, setOpenItems] = useState<string[]>([])
+  const [floatingIpModalOpen, setFloatingIpModalOpen] = useState(false)
+  const [selectedFloatingIp, setSelectedFloatingIp] = useState<FloatingIp | undefined>()
   const externalIps = useController({ control, name: 'externalIps' })
   const ephemeralIp = externalIps.field.value?.find((ip) => ip.type === 'ephemeral')
   const assignEphemeralIp = !!ephemeralIp
   const selectedPool = ephemeralIp && 'pool' in ephemeralIp ? ephemeralIp.pool : undefined
   const defaultPool = siloPools.find((pool) => pool.isDefault)?.name
+  const attachedFloatingIps = (externalIps.field.value || []).filter(isFloating)
+
+  // Filter out the IPs that are already attached to an instance
+  const attachableFloatingIps = useMemo(
+    () => floatingIpList.items.filter((ip) => !ip.instanceId),
+    [floatingIpList]
+  )
+
+  // To find available floating IPs, we remove the ones that are already committed to this instance
+  const availableFloatingIps = attachableFloatingIps.filter(
+    (ip) => !attachedFloatingIps.find((attachedIp) => attachedIp.floatingIp === ip.name)
+  )
+  const attachedFloatingIpsData = attachedFloatingIps
+    .map((ip) => attachableFloatingIps.find((fip) => fip.name === ip.floatingIp))
+    .filter((x) => !!x) as Array<FloatingIp>
+
+  const closeFloatingIpModal = () => {
+    setFloatingIpModalOpen(false)
+    setSelectedFloatingIp(undefined)
+  }
+
+  const attachFloatingIp = () => {
+    externalIps.field.onChange([
+      ...(externalIps.field.value || []),
+      { type: 'floating', floatingIp: selectedFloatingIp?.name },
```

Comment:
I was surprised this typechecks because `floatingIp` is a required string, but the value here could be undefined (a possibility we can eliminate by wrapping the inside of the function in `if (selectedFloatingIp) { }`).

<img width="596" alt="image" src="https://github.com/oxidecomputer/console/assets/3612203/79d61667-d021-4bf4-925e-94a642b9ab80">

It looks like `onChange` is not intelligently typed even though `field` is:

<img width="599" alt="image" src="https://github.com/oxidecomputer/console/assets/3612203/734f0dd8-e10b-4c5a-85f2-9d8231d41527">


<img width="787" alt="image" src="https://github.com/oxidecomputer/console/assets/3612203/05644773-d0c6-4689-9719-c9dfdb1ae174">

That's kind of shitty! It' s not generic either, so it doesn't seem like there's anything we can do about it:

https://github.com/react-hook-form/react-hook-form/blob/927535bf87412b2baa91faefb4c7ded327126c2b/src/types/controller.ts#L27